platform/mellanox/mlnx-platform-api/sonic_platform/component.py

```diff
@@ -132,17 +248,86 @@ def get_firmware_version(self):
 
         return cpld_version
 
+
+    def _get_mst_device(self):
+        mst_dev_list = glob(MST_DEVICE_PATTERN)
+        if mst_dev_list is None or len(mst_dev_list) != 1:
+            return None
+        return mst_dev_list
+
+
+    def install_firmware(self, image_path):
+        """
+        Installs firmware to the component
+
+        Args:
+            image_path: A string, path to firmware image
+
+        Returns:
+            A boolean, True if install was successful, False if not
+
+        Details:
+            The command "cpldupdate" is provided to install CPLD. There are two ways to do it:
+                1. To burn CPLD via gpio, which is faster but only supported on new systems, like Anaconda, ...
+                2. To install CPLD via firmware, which is slower but supported on older systems.
+                   This also requires the mst device designated.
+            "cpldupdate --dev <devname> <vme_file>" has the logic of testing whether to update via gpio is supported,
+            and if so then go this way, otherwise tries updating software via fw. So we take advantage of it to update the CPLD. 
+            By doing so we don't have to mind whether to update via gpio supported, which belongs to hardware details.
+
+            So the procedure should be:
+                1. Test whether the file exists
+                2. Fetch the mst device name
+                3. Update CPLD via executing "cpldupdate --dev <devname> <vme_file>"
+                4. Check the result
+        """
+        # check whether the image file exists
+        if not self._check_file_validity(image_path):
+            return False
+
+        mst_dev_list = self._get_mst_device()
+        if mst_dev_list is None:
+            print("Failed to get mst device which is required for CPLD updating or multiple device files matched")
+            return False
+
+        cmdline = self.CPLD_UPDATE_COMMAND.format(mst_dev_list[0], image_path)
+        outputline = ""
+        success_flag = False
+        try:
+            proc = subprocess.Popen(cmdline, stdout=subprocess.PIPE, shell=True, stderr=subprocess.STDOUT)
+            while True:
+                out = proc.stdout.read(1)
+                if out == '' and proc.poll() != None:
+                    break
+                if out != '':
+                    sys.stdout.write(out)
+                    sys.stdout.flush()
+                    outputline += out
+                if (out == '\n' or out == '\r') and len(outputline):
+                    m = re.search(self.CPLD_INSTALL_SUCCESS_FLAG, outputline)
+                    if m and m.group(0) == self.CPLD_INSTALL_SUCCESS_FLAG:
+                        success_flag = True
+
+        except OSError as e:
+            raise RuntimeError("Failed to execute command {} due to {}".format(cmdline, repr(e)))
+
+        if success_flag:
+            print("Success. Refresh or power cycle is required to finish CPLD installation.")
```

Comment:
@stephenxs what about some simple logger for stuff like this?
For example:
```
INFO: Refresh or power cycle is required to finish CPLD installation
ERROR: Failed to install CPLD
```

Reply:
fixed.